dynacred/src/Contact.ts

```diff
@@ -123,82 +135,98 @@ export class Contact {
     }
 
     set phone(p: string) {
-        if (p) {
+        if (this.phone !== p) {
             this.credential.setAttribute("1-public", "phone", Buffer.from(p));
             this.incVersion();
         }
     }
 
-    get seedPublic(): Public {
-        return Public.fromBuffer(this.credential.getAttribute("1-public", "seedPub"));
+    get seedPublic(): Public | undefined {
+        const raw = this.credential.getAttribute("1-public", "seedPub");
+        if (raw === undefined) {
+            return undefined;
+        }
+        return Public.fromBuffer(raw);
     }
 
     set seedPublic(pub: Public) {
-        if (pub) {
+        if (this.seedPublic === undefined || !this.seedPublic.equal(pub)) {
             this.credential.setAttribute("1-public", "seedPub", pub.toBuffer());
             this.incVersion();
         }
     }
 
+    get joinedChallenge(): Long {
+        const longArray = this.credential.getAttribute("1-public", "challenge");
+        return Long.fromBytesLE(Array.from(longArray));
+    }
+
+    set joinedChallenge(v: Long) {
+        if (!this.joinedChallenge.equals(v)) {
+            this.credential.setAttribute("1-public", "challenge", Buffer.from(v.toBytesLE()));
+            this.incVersion();
+        }
+    }
+
     get personhoodPub(): Public | undefined {
         const raw = this.credential.getAttribute("1-public", "personhood");
         if (raw === undefined) {
             return undefined;
         }
-        if (raw === null) {
-            Log.warn(`Contact("${this.alias}").credential has null 1-public/personhood`);
-            return undefined;
-        }
         return Public.fromBuffer(raw);
     }
 
     set personhoodPub(pub: Public) {
-        if (pub) {
+        if (this.personhoodPub === undefined ||
```

Comment:
`undefined` checked by `pub.equal`, you can remove it.

Reply:
I'm not sure that `pub.equals(undefined)` actually works...

Reply:
You're right, it doesn't support `undefined`.